django/http/response.py

```diff
@@ -387,15 +389,42 @@ class FileResponse(StreamingHttpResponse):
     """
     block_size = 4096
 
+    def __init__(self, *args, as_attachment=False, filename='', **kwargs):
+        self.as_attachment = as_attachment
+        self.filename = filename
+        super().__init__(*args, **kwargs)
+
     def _set_streaming_content(self, value):
-        if hasattr(value, 'read'):
-            self.file_to_stream = value
-            filelike = value
-            if hasattr(filelike, 'close'):
-                self._closable_objects.append(filelike)
-            value = iter(lambda: filelike.read(self.block_size), b'')
-        else:
+        if not hasattr(value, 'read'):
             self.file_to_stream = None
+            return super()._set_streaming_content(value)
+
+        self.file_to_stream = value
+        filelike = value
+        if hasattr(filelike, 'close'):
+            self._closable_objects.append(filelike)
+        value = iter(lambda: filelike.read(self.block_size), b'')
+
+        if isinstance(getattr(filelike, 'name', None), str) and os.path.isabs(filelike.name):
+            self['Content-Length'] = os.path.getsize(filelike.name)
+            if self.get('Content-Type', '').startswith(settings.DEFAULT_CONTENT_TYPE):
+                content_type, encoding = mimetypes.guess_type(filelike.name)
+                self['Content-Type'] = content_type or 'application/octet-stream'
+                if encoding:
+                    self['Content-Encoding'] = encoding
```

Comment:
For a gzip file, I believe this will cause the browser to unzip it.

```
$ python3
>>> import mimetypes
>>> mimetypes.guess_type('f.tar.gz')
('application/x-tar', 'gzip')
```

But I'm guessing the user wants the mime type to be `application/gzip` (or similar) with no encoding.

[MDN Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)

> The Content-Encoding entity header is used to compress the media-type. When present, its value indicates which encodings were applied to the entity-body. It lets the client know, how to decode in order to obtain the media-type referenced by the Content-Type header.

I know of at least one third party Django app that has this issue: https://github.com/johnsensible/django-sendfile/issues/38